coderd/database/pubsub/pubsub.go

```diff
@@ -174,29 +175,38 @@ type pgPubsub struct {
 	queues           map[string]map[uuid.UUID]*msgQueue
 	closedListener   bool
 	closeListenerErr error
+
+	publishesTotal      *prometheus.CounterVec
+	subscribesTotal     *prometheus.CounterVec
+	messagesTotal       *prometheus.CounterVec
+	disconnectionsTotal prometheus.Counter
+	connected           prometheus.Gauge
 }
 
 // BufferSize is the maximum number of unhandled messages we will buffer
 // for a subscriber before dropping messages.
 const BufferSize = 2048
 
 // Subscribe calls the listener when an event matching the name is received.
-func (p *pgPubsub) Subscribe(event string, listener Listener) (cancel func(), err error) {
+func (p *PGPubsub) Subscribe(event string, listener Listener) (cancel func(), err error) {
 	return p.subscribeQueue(event, newMsgQueue(p.ctx, listener, nil))
 }
 
-func (p *pgPubsub) SubscribeWithErr(event string, listener ListenerWithErr) (cancel func(), err error) {
+func (p *PGPubsub) SubscribeWithErr(event string, listener ListenerWithErr) (cancel func(), err error) {
 	return p.subscribeQueue(event, newMsgQueue(p.ctx, nil, listener))
 }
 
-func (p *pgPubsub) subscribeQueue(event string, newQ *msgQueue) (cancel func(), err error) {
+func (p *PGPubsub) subscribeQueue(event string, newQ *msgQueue) (cancel func(), err error) {
 	p.mut.Lock()
 	defer p.mut.Unlock()
 	defer func() {
 		if err != nil {
 			// if we hit an error, we need to close the queue so we don't
 			// leak its goroutine.
 			newQ.close()
+			p.subscribesTotal.WithLabelValues("false").Inc()
+		} else {
+			p.subscribesTotal.WithLabelValues("true").Inc()
```

Comment:
How about including the event as well? Might be useful to see if there's an abnormal amount of subs/pubs/data going for a specific event.

Reply:
That's a bad idea because some events include the UUID of the thing, e.g. workspace, which would make the set of label values unbounded, which is a cardinal sin in metrics.

Not for this PR, but if we decide we really want that information, we could standardize event channel names to be a struct with a static root and then zero or more specific IDs, and then we could label the publishes, subscribes and messages with the static root.